examples/sparse/gcnii.py

```diff
@@ -0,0 +1,179 @@
+import math
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity, spmm
+from torch.nn.parameter import Parameter
+from torch.optim import Adam
+
+
+class GCNIIConvolution(nn.Module):
+    def __init__(self, in_size, out_size):
+        super().__init__()
+        self.out_size = out_size
+        self.weight = Parameter(torch.FloatTensor(in_size, out_size))
+        self.reset_parameters()
+
+    ############################################################################
+    # (HIGHLIGHT) Take the advantage of DGL sparse APIs to implement the GCNII
+    # forward process.
+    ############################################################################
+    def forward(self, A_norm, H, H0, lamda, alpha, l):
+        beta = math.log(lamda / l + 1)
+        H = spmm(A_norm, H)
+        support = (1 - alpha) * H + alpha * H0
+        H = (1 - beta) * support + beta * support @ self.weight
+        return H
+
+    def reset_parameters(self):
+        std = 1 / math.sqrt(self.out_size)
+        self.weight.data.uniform_(-std, std)
+
+
+class GCNII(nn.Module):
+    def __init__(
+        self,
+        in_size,
+        out_size,
+        hidden_size,
+        n_layers,
+        lamda,
+        alpha,
+        dropout=0.5,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.n_layers = n_layers
+        self.lamda = lamda
+        self.alpha = alpha
+
+        self.FC_layers = nn.ModuleList()
+        self.FC_layers.append(nn.Linear(in_size, hidden_size))
+        self.FC_layers.append(nn.Linear(hidden_size, out_size))
+
+        self.CONV_layers = nn.ModuleList()
+        for _ in range(n_layers):
+            self.CONV_layers.append(GCNIIConvolution(hidden_size, hidden_size))
+        self.activation = nn.ReLU()
+        self.dropout = dropout
+
+    def forward(self, A_norm, feature):
+        H = feature
+        H = F.dropout(H, self.dropout)
+        H = self.FC_layers[0](H)
+        H = self.activation(H)
+        H0 = H
+        for i, conv in enumerate(self.CONV_layers):
+            H = F.dropout(H, self.dropout)
+            H = conv(A_norm, H, H0, self.lamda, self.alpha, i + 1)
+            H = self.activation(H)
+        H = F.dropout(H, self.dropout)
+        H = self.FC_layers[-1](H)
+
+        return F.log_softmax(H, dim=1)
+
+
+def evaluate(g, pred):
+    label = g.ndata["label"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+
+    # Compute accuracy on validation/test set.
+    val_acc = (pred[val_mask] == label[val_mask]).float().mean()
+    test_acc = (pred[test_mask] == label[test_mask]).float().mean()
+    return val_acc, test_acc
+
+
+def train(args, model, g, A_norm, H):
+    label = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    optimizer = Adam(model.parameters(), args.lr, weight_decay=args.wd)
+
+    loss_fcn = nn.NLLLoss()
```

Comment:
For consistency with other examples, use `CrossEntropyLoss` with logits?

Reply:
use CrossEntropyLoss instead